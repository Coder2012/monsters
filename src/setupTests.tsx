// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom/extend-expect';
import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import type { PreloadedState } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/dist/query';
import { Provider } from 'react-redux';

import { AppStore, RootState, setupStore } from './app/store';
import { server } from './mocks/server';
import { apiSlice } from './features/api/apiSlice';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  setupListeners(store.dispatch);

  const wrapper = ({ children }: PropsWithChildren<{}>): JSX.Element => (
    <Provider store={store}>{children}</Provider>
  );
  return { store, ...render(ui, { wrapper, ...renderOptions }) };
}

const store = setupStore({} as PreloadedState<RootState>);

// Establish API mocking before all tests.
beforeAll(() => {
  server.listen();
});

// Reset any request handlers that we may add during the tests,
// so they don't affect other tests.
afterEach(() => {
  server.resetHandlers();
  // This is the solution to clear RTK Query cache after each test
  store.dispatch(apiSlice.util.resetApiState());
});

// Clean up after the tests are finished.
afterAll(() => server.close());
