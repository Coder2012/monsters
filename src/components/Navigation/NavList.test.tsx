import { MemoryRouter } from 'react-router-dom';
import { renderWithProviders } from '../../setupTests';
import { NavList } from './NavList';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  __esModule: true,
  useLocation: jest.fn(),
}));

describe('NavList', () => {
  it.only('should render navigation', () => {
    const { container } = renderWithProviders(
      <MemoryRouter initialEntries={['/kids/']}>
        <NavList />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
