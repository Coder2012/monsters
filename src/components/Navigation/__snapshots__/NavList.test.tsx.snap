// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavList should render navigation 1`] = `
<div>
  <nav
    class="container"
  >
    <ul>
      <li>
        <a
          class="link"
          href="/"
        >
          Monsters
        </a>
      </li>
      <li>
        <a
          aria-current="page"
          class="active"
          href="/kids"
        >
          Kids
        </a>
      </li>
      <li>
        <a
          class="link"
          href="/tasks"
        >
          Tasks
        </a>
      </li>
      <li>
        <a
          class="link"
          href="/leaderboard"
        >
          Leaderboard
        </a>
      </li>
    </ul>
  </nav>
</div>
`;
